build: add remark plugin to validate expected HTML sections

Planeshifter opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   adds a remark plugin to ensure that `README.md` files have the expected HTML `<section>` tag structure as mandated by stdlib conventions.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses https://github.com/stdlib-js/metr-issue-tracker/issues/49 by resolving 3).

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Planeshifter committed:
build: add remark plugin to validate expected HTML sections

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
docs: add missing require to example code

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
build: add options validation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-expected-html-sections/lib/linter.js:
```diff
@@ -0,0 +1,184 @@
+			// Check if this is a section start tag:
+			sectionMatch = SECTION_START.exec( node.value );
+			if ( sectionMatch ) {
+				className = sectionMatch[1] || '';
```
Why is `className` declared in the parent scope?

kgryte replied:
In general, can the `visitor` function be moved to a parent scope?

Planeshifter replied:
Moved variable declarations to `visitor` scope as appropriate. However, we cannot move the `visitor` function easily to parent scope since it uses several variables from the linter function scope (`sectionsFound`, `sectionStack`, ...)

Planeshifter committed:
chore: move declarations to visitor function and other clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-expected-html-sections/lib/linter.js:
```diff
@@ -0,0 +1,186 @@
+		}
+
+		if ( missingRoot.length > 0 ) {
+			msg = 'Missing required root-level sections: `' + missingRoot.join( '`, `' ) + '`. Required sections are: `' + requiredRootSections.join('`, `') + '`.   missing-required-sections';
```
```suggestion
			msg = 'Missing required root-level sections: `' + missingRoot.join( '`, `' ) + '`. Required sections are: `' + requiredRootSections.join( '`, `' ) + '`.   missing-required-sections';
```

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-expected-html-sections/lib/linter.js:
```diff
@@ -0,0 +1,186 @@
+
+		// If 'c' section exists, check its requirements:
+		if ( sectionsFound.root.c ) {
+			requiredCSections = (schema.c) ? (schema.c.required || []) : [];
```
```suggestion
			requiredCSections = ( schema.c ) ? ( schema.c.required || [] ) : [];
```

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-expected-html-sections/lib/linter.js:
```diff
@@ -0,0 +1,186 @@
+			}
+
+			if ( missingC.length > 0 ) {
+				msg = 'Missing required sections in "c" section: `' + missingC.join('`, `') + '`. Required C sections are: `' + requiredCSections.join('`, `') + '`.   missing-required-c-sections';
```
```suggestion
				msg = 'Missing required sections in "c" section: `' + missingC.join( '`, `' ) + '`. Required C sections are: `' + requiredCSections.join( '`, `' ) + '`.   missing-required-c-sections';
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-expected-html-sections/lib/linter.js:
```diff
@@ -0,0 +1,186 @@
+			var className;
+
+			// Check if this is a section start tag:
+			sectionMatch = RE_SECTION_START.exec( node.value );
```
Similar question regarding the use of `exec`.

Planeshifter replied:
No global regular expressions, so no statefulness here and not a concern (see: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec#description).

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-expected-html-sections/lib/linter.js:
```diff
@@ -0,0 +1,186 @@
+				if ( sectionStack.length > 0 ) {
+					sectionStack.pop();
+				}
+			}
```
```suggestion
			else if ( RE_SECTION_END.test( node.value ) && sectionStack.length > 0 ) {
				sectionStack.pop();
			}
```

You can refactor to avoid an extra level of nesting.

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-expected-html-sections/lib/validate.js:
```diff
@@ -0,0 +1,70 @@
+		return new TypeError( format( 'invalid argument. Options argument must be an object. Value: `%s`.', options ) );
+	}
+	if ( hasOwnProp( options, 'schema' ) ) {
+		if ( !isObject( options.schema ) ) {
```
This is a fairly weak check, as I could pass in `options.schema = { 'foo': 'bar', 'required': null }` and this validation test would pass.

Planeshifter replied:
Added full validation for the passed schema.

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-expected-html-sections/test/test.js:
```diff
@@ -0,0 +1,165 @@
+			t.fail( err.message );
+		}
+		t.strictEqual( file.messages.length, 1, 'has one error for invalid asin fixture' );
+		t.strictEqual( file.messages[0].reason.includes('Missing required sections in "c" section: `usage`'), true, 'error identifies the missing C usage section' );
```
Refactor to avoid use of `includes` here and below. Also, spacing.

Planeshifter replied:
Addressed.

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-expected-html-sections/test/test.js:
```diff
@@ -0,0 +1,165 @@
+			t.fail( err.message );
+		}
+		t.strictEqual( file.messages.length, 1, 'has one error for fixture missing a required root section' );
+		t.strictEqual( file.messages[0].reason.includes('Missing required root-level sections: `usage`'), true, 'error identifies the missing section' );
```
Same comment.

Planeshifter replied:
Addressed.

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-expected-html-sections/test/test.js:
```diff
@@ -0,0 +1,165 @@
+			t.fail( err.message );
+		}
+		t.strictEqual( file.messages.length, 1, 'has one error for fixture with an incomplete C section' );
+		t.strictEqual( file.messages[0].reason.includes('Missing required sections in "c" section: `usage`, `examples`'), true, 'error identifies the missing C sections' );
```
Same comment.

Planeshifter replied:
Addressed.

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-expected-html-sections/test/test.js:
```diff
@@ -0,0 +1,165 @@
+			t.fail( err.message );
+		}
+		t.strictEqual( file.messages.length, 1, 'has correct number of errors with custom schema' );
+		t.strictEqual( file.messages[0].reason.includes('usage') && file.messages[0].reason.includes('related'), true, 'error identifies both missing sections' );
```
Same comment.

Planeshifter replied:
Addressed.

kgryte reviewed (commented):
Left an initial round of comments.

Planeshifter committed:
chore: apply suggestions from code review

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
chore: add validation of schema

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-expected-html-sections/lib/validate.js:
```diff
@@ -0,0 +1,142 @@
+		}
+		opts.schema = options.schema;
+	}
+
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-expected-html-sections/lib/validate.js:
```diff
@@ -0,0 +1,142 @@
+	if ( !isObject( schema ) ) {
+		return new TypeError( format( 'invalid option. `%s` schema section must be an object. Value: `%s`.', section, schema ) );
+	}
+
+	// Check if 'required' exists and is an array of strings
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-expected-html-sections/lib/validate.js:
```diff
@@ -0,0 +1,142 @@
+		if ( !isArray( schema.required ) ) {
+			return new TypeError( format( 'invalid option. `%s.required` must be an array. Value: `%s`.', section, schema.required ) );
+		}
+		// Validate each element in the 'required' array is a string
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-expected-html-sections/lib/validate.js:
```diff
@@ -0,0 +1,142 @@
+	} else {
+		return new TypeError( format( 'invalid option. `%s` schema section must have a `required` property.', section ) );
+	}
+
+	// Check if 'optional' exists and is an array of strings
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-expected-html-sections/lib/validate.js:
```diff
@@ -0,0 +1,142 @@
+		if ( !isArray( schema.optional ) ) {
+			return new TypeError( format( 'invalid option. `%s.optional` must be an array. Value: `%s`.', section, schema.optional ) );
+		}
+		// Validate each element in the 'optional' array is a string
```
```suggestion
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-expected-html-sections/lib/validate.js:
```diff
@@ -0,0 +1,135 @@
+	} else {
+		return new TypeError( format( 'invalid option. `%s` schema section must have an `optional` property.', section ) );
+	}
+
```
```suggestion
```

kgryte committed:
docs: remove empty line

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-expected-html-sections/lib/validate.js:
```diff
@@ -0,0 +1,134 @@
+var isObject = require( '@stdlib/assert/is-plain-object' );
+var hasOwnProp = require( '@stdlib/assert/has-own-property' );
+var isArray = require( '@stdlib/assert/is-array' );
+var isString = require( '@stdlib/assert/is-string' );
```
```suggestion
var isString = require( '@stdlib/assert/is-string' ).isPrimitive;
```

kgryte committed:
refactor: restrict allowed values to primitives

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
chore: clean-up

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
chore: clean-up

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-expected-html-sections/test/test.js:
```diff
@@ -0,0 +1,166 @@
+		'optional': [ 'intro', 'notes', 'c', 'references' ]
+	}
+};
+var FIXTURE_VALID_COMPLETE = readFileSync( join( TEST_FIXTURES, 'valid_complete.md.txt' ), 'utf8' ).toString();
+var FIXTURE_VALID_MEMOIZE = readFileSync( join( TEST_FIXTURES, 'valid_memoize.md.txt' ), 'utf8' ).toString();
+var FIXTURE_VALID_ASIN = readFileSync( join( TEST_FIXTURES, 'valid_asin.md.txt' ), 'utf8' ).toString();
+var FIXTURE_INVALID_ASIN = readFileSync( join( TEST_FIXTURES, 'invalid_asin.md.txt' ), 'utf8' ).toString();
+var FIXTURE_MISSING_ROOT = readFileSync( join( TEST_FIXTURES, 'missing_required_root.md.txt' ), 'utf8' ).toString();
+var FIXTURE_INCOMPLETE_C = readFileSync( join( TEST_FIXTURES, 'incomplete_c.md.txt' ), 'utf8' ).toString();
```
```suggestion
var opts = {
	'encoding': 'utf8'
};
var FIXTURE_VALID_COMPLETE = readFileSync( join( TEST_FIXTURES, 'valid_complete.md.txt' ), opts );
var FIXTURE_VALID_MEMOIZE = readFileSync( join( TEST_FIXTURES, 'valid_memoize.md.txt' ), opts );
var FIXTURE_VALID_ASIN = readFileSync( join( TEST_FIXTURES, 'valid_asin.md.txt' ), opts );
var FIXTURE_INVALID_ASIN = readFileSync( join( TEST_FIXTURES, 'invalid_asin.md.txt' ), opts );
var FIXTURE_MISSING_ROOT = readFileSync( join( TEST_FIXTURES, 'missing_required_root.md.txt' ), opts );
var FIXTURE_INCOMPLETE_C = readFileSync( join( TEST_FIXTURES, 'incomplete_c.md.txt' ), opts );
```

kgryte committed:
chore: clean-up

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
refactor: consolidate error messages

Signed-off-by: Athan <kgryte@gmail.com>
